### PR TITLE
Add get_checkpoint_block to the fork choice spec

### PR DESCRIPTION
**Pull request by saltiniroberto**
This PR introduces the function `get_checkpoint_block` and replaces any computation of checkpoint blocks in the fork choice spec with a call to the newly introduced function.

The objective of this PR is to increase the readability of the spec by better clarifying when we are interested in calculating an checkpoint block.

### CODE REVIEW

**Review by ralexstokes — COMMENTED** [review]
looks good just from a refactor perspective and I like the increased semantic content

**Review by hwwhww — COMMENTED** [review]
It looks like a nice-to-have to me. No strong opinion here.

/cc @adiasg 

Some CircleCI didn't start with the latest commit. It may need a kick with a new commit. 😬

File: specs/phase0/fork-choice.md [hunk 1]
```diff
@@ -506,7 +516,7 @@ def on_block(store: Store, signed_block: SignedBeaconBlock) -> None:
     finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
     assert block.slot > finalized_slot
     # Check block is a descendant of the finalized block at the checkpoint finalized slot
-    assert get_ancestor(store, block.parent_root, finalized_slot) == store.finalized_checkpoint.root
```
[the remark this answers is not in the document]
Comment (saltiniroberto) [reply specs/phase0/fork-choice.md:509]: @hwwhww Why do you think that there may be no need to modify this one?

File: specs/phase0/fork-choice.md [hunk 2]
```diff
@@ -192,6 +193,17 @@ def get_ancestor(store: Store, root: Root, slot: Slot) -> Root:
     return root
 ```
 
+#### `get_epoch_boundary_block`
+
+```python
+def get_epoch_boundary_block(store: Store, root: Root, epoch: Epoch) -> Root:
```
[the remark this answers is not in the document]
Comment (saltiniroberto) [reply specs/phase0/fork-choice.md:199]: `get_ancestor_at_epoch_boundary` seems a good description to me

File: specs/phase0/fork-choice.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (saltiniroberto) [reply specs/phase0/fork-choice.md:509]: The objective of this PR is not so much about simplifying the code, but rather making explicit when we are computing an epoch boundary block."

For example, in this line, we are not only checking that `store.finalized_checkpoint.root` is an ancestor of `block.parent_root`, but also that `store.finalized_checkpoint.root` is an epoch boundary block in the chain of `block.parent_root`, which I don't think that it is immediate in the original code.

[the remark this answers is not in the document]
Comment (saltiniroberto) [reply specs/phase0/fork-choice.md:509]: But the difference between checking whether a block is just an ancestor or it is also an epoch boundary block is significant when performing security analysis on the code.

**Review by mkalinin — APPROVED** [review]

**Review by potuz — CHANGES_REQUESTED** [review]
You need to change the corresponding files in different forks that changed `on_block`, eg. `Bellatrix`

**Comment by saltiniroberto** [comment]
> You need to change the corresponding files in different forks that changed `on_block`, eg. `Bellatrix`
:+1: 
Done.

**Comment by ppopth** [comment]
Please change the p2p network specs as well.
```
specs/phase0/p2p-interface.md:  `get_ancestor(store, block.parent_root, compute_start_slot_at_epoch(store.finalized_checkpoint.epoch))
specs/phase0/p2p-interface.md:  `get_ancestor(store, aggregate.data.beacon_block_root, compute_start_slot_at_epoch(store.finalized_checkpoint.epoch))
specs/phase0/p2p-interface.md:  `get_ancestor(store, attestation.data.beacon_block_root, compute_start_slot_at_epoch(attestation.data.target.epoch)) == attestation.data.target.root`
specs/phase0/p2p-interface.md:  `get_ancestor(store, attestation.data.beacon_block_root, compute_start_slot_at_epoch(store.finalized_checkpoint.epoch))
```

**Comment by ppopth** [comment]
We should also update the pyspec correspondingly, if we want to apply this patch.

For example, in `tests/core/pyspec/eth2spec/test/phase0/fork_choice/test_on_block.py`, we should change the following.
```python
finalized_slot = spec.compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
ancestor_at_finalized_slot = spec.get_ancestor(store, pre_store_justified_checkpoint_root, finalized_slot)
```

Please run `git grep get_ancestor tests/` to see all the places that need a change.

File: specs/deneb/fork-choice.md [hunk 3]
```diff
@@ -82,7 +82,11 @@ def on_block(store: Store, signed_block: SignedBeaconBlock) -> None:
     finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
     assert block.slot > finalized_slot
     # Check block is a descendant of the finalized block at the checkpoint finalized slot
-    assert get_ancestor(store, block.parent_root, finalized_slot) == store.finalized_checkpoint.root
+    assert store.finalized_checkpoint.root == get_ancestor_at_epoch_boundary(
+        store,
+        block.parent_root,
+        store.finalized_checkpoint.epoch,
+    )
```
Comment (djrtwo) [inline specs/deneb/fork-choice.md:89]: Doing an assignment before the assert is preferable to reduce line count and increase readability. ditto at a few other places in this patch

```suggestion
    epoch_boundary_root = get_ancestor_at_epoch_boundary(store, block.parent_root, store.finalized_checkpoint.epoch)
    assert store.finalized_checkpoint.root == epoch_boundary_root
```

File: specs/phase0/fork-choice.md [hunk 4]
```diff
@@ -278,10 +290,13 @@ def filter_block_tree(store: Store, block_root: Root, blocks: Dict[Root, BeaconB
             voting_source.epoch + 2 >= current_epoch
         )
 
-    finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
     correct_finalized = (
         store.finalized_checkpoint.epoch == GENESIS_EPOCH
-        or store.finalized_checkpoint.root == get_ancestor(store, block_root, finalized_slot)
+        or store.finalized_checkpoint.root == get_ancestor_at_epoch_boundary(
```
Comment (djrtwo) [inline specs/phase0/fork-choice.md:295]: ditto

File: specs/phase0/fork-choice.md [hunk 5]
```diff
@@ -506,7 +520,11 @@ def on_block(store: Store, signed_block: SignedBeaconBlock) -> None:
     finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
     assert block.slot > finalized_slot
     # Check block is a descendant of the finalized block at the checkpoint finalized slot
-    assert get_ancestor(store, block.parent_root, finalized_slot) == store.finalized_checkpoint.root
+    assert store.finalized_checkpoint.root == get_ancestor_at_epoch_boundary(
```
Comment (djrtwo) [inline specs/phase0/fork-choice.md:523]: ditto

**Review by adiasg — CHANGES_REQUESTED** [review]
The fork choice & p2p spec do not refer to epoch boundary blocks before this PR. More generally, the spec refers to epoch boundary blocks as part of `Checkpoint`s. 

I'd prefer to avoid introducing the "epoch boundary block" nomenclature here, and instead rename `get_ancestor_at_epoch_boundary` to `get_checkpoint_block`.

File: specs/deneb/fork-choice.md (hunk 3 above)
Comment (saltiniroberto) [reply specs/deneb/fork-choice.md]: Done here in all other places marked with "ditto"

**Comment by saltiniroberto** [comment]
`make test` report was

> ================================================================================================================================================= short test summary info ==================================================================================================================================================
FAILED eth2spec/test/deneb/unittests/polynomial_commitments/test_polynomial_commitments.py::test_validate_kzg_g1_not_in_g1 - AssertionError: expected an assertion error, but got none.
FAILED eth2spec/test/deneb/unittests/polynomial_commitments/test_polynomial_commitments.py::test_validate_kzg_g1_not_on_curve - AssertionError: expected an assertion error, but got none.
================================================================================================================================= 2 failed, 806 passed, 146 skipped in 2537.18s (0:42:17) ==================================================================================================================================

**Comment by saltiniroberto** [comment]
> The fork choice & p2p spec do not refer to epoch boundary blocks before this PR. More generally, the spec refers to epoch boundary blocks as part of `Checkpoint`s.
> 
> I'd prefer to avoid introducing the "epoch boundary block" nomenclature here, and instead rename `get_ancestor_at_epoch_boundary` to `get_checkpoint_block`.

Done

**Review by adiasg — APPROVED** [review]

**Pull request merged**